command/agent/template/template.go

```diff
@@ -0,0 +1,186 @@
+		// nothing to do
+		close(ts.unblockCh)
+		return
+	}
+
+	// construct a consul template vault config based the agents vault
+	// configuration
+	var runnerConfig *ctconfig.Config
+	if runnerConfig = newRunnerConfig(ts.config, templates); runnerConfig == nil {
+		ts.logger.Info("template server failed to generate runner config")
+		return
+	}
+
+	var err error
+	ts.runner, err = manager.NewRunner(runnerConfig, false)
+	if err != nil {
+		ts.logger.Info("template server failed to create")
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			ts.runner.StopImmediately()
+			return
+
+		case token := <-incoming:
+			if token != *latestToken {
+				ts.logger.Info("template server received new token")
+				ts.runner.Stop()
+				*latestToken = token
+				ctv := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token: latestToken,
+					},
+				}
+				runnerConfig.Merge(&ctv)
+				runnerConfig.Finalize()
+				var runnerErr error
+				ts.runner, runnerErr = manager.NewRunner(runnerConfig, false)
+				if runnerErr != nil {
+					ts.logger.Info("template server failed with new Vault token")
+					// TODO: continue or fail here? Right now we just continue
+					continue
+				} else {
+					go ts.runner.Start()
+				}
+			}
+		case err := <-ts.runner.ErrCh:
+			ts.logger.Info("template server error:", err.Error())
+			return
+		case <-ts.runner.TemplateRenderedCh():
+			// A template has been rendered, unblock
+			if ts.exitAfterAuth {
+				// if we want to exit after auth, go ahead and shut down the runner
+				ts.runner.Stop()
+			}
+			close(ts.unblockCh)
+		}
+	}
+}
+
+// newRunnerConfig returns a consul-template runner configuration, setting the
+// Vault and Consul configurations based on the clients configs.
+func newRunnerConfig(sc *ServerConfig, templates ctconfig.TemplateConfigs) *ctconfig.Config {
+	// TODO only use default Vault config
+	conf := ctconfig.DefaultConfig()
+	conf.Templates = templates.Copy()
+
+	// Setup the Vault config
+	// Always set these to ensure nothing is picked up from the environment
+	emptyStr := ""
+	conf.Vault.RenewToken = boolPtr(false)
+	conf.Vault.Token = &emptyStr
+	conf.Vault.Address = &sc.VaultConf.Address
+
+	if sc.Namespace != "" {
+		conf.Vault.Namespace = &sc.Namespace
+	}
+
+	conf.Vault.SSL = &ctconfig.SSLConfig{
+		Enabled:    boolPtr(false),
+		Verify:     boolPtr(false),
+		Cert:       &emptyStr,
+		Key:        &emptyStr,
+		CaCert:     &emptyStr,
+		CaPath:     &emptyStr,
+		ServerName: &emptyStr,
+	}
+
+	if strings.HasPrefix(sc.VaultConf.Address, "https") || sc.VaultConf.CACert != "" {
+		skipVerify := sc.VaultConf.TLSSkipVerify
+		verify := !skipVerify
+		conf.Vault.SSL = &ctconfig.SSLConfig{
+			Enabled: boolPtr(true),
+			Verify:  &verify,
+			Cert:    &sc.VaultConf.ClientCert,
+			Key:     &sc.VaultConf.ClientKey,
+			CaCert:  &sc.VaultConf.CACert,
+			CaPath:  &sc.VaultConf.CAPath,
+		}
+	}
+
+	conf.Finalize()
+	return conf
+}
+
+func boolPtr(b bool) *bool {
```

Comment:
This is also one of those funcs that would be pretty cool to have in the sdk helpers.

Reply:
I do agree. I don't know I'll do that in this feature branch though. I'm pretty sure this is the only `Ptr` method I actually use/introduce that is in the normal code and not in test code, so I think for that reason I just left it. 

IF you feel strongly that this should be done let me know, otherwise I'll defer

Reply:
That's fine, it's only used by this code at the present. Also I don't want to make this PR harder to get through due to editing common code.